# Process_data : 

In [90]:
import pandas as pd
import os
import json

In [91]:
path_energy = os.path.join("global-data-sustainable-energy.csv")
path_cc = os.path.join("countryContinent.csv")
path_hdi = os.path.join("human-development-index.csv")
path_wp = os.path.join("worldPolygone.geo.json")

## Lecture et Analyse globale du jeu de donnée

In [92]:
df_energy = pd.read_csv(path_energy)
df_energy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3649 entries, 0 to 3648
Data columns (total 21 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            3649 non-null   object 
 1   Year                                                              3649 non-null   int64  
 2   Access to electricity (% of population)                           3639 non-null   float64
 3   Access to clean fuels for cooking                                 3480 non-null   float64
 4   Renewable-electricity-generating-capacity-per-capita              2718 non-null   float64
 5   Financial flows to developing countries (US $)                    1560 non-null   float64
 6   Renewable energy share in the total final energy consumption (%)  3455 non-null   float64
 7   Electricity from fossil fuels (TW

In [93]:
df_energy.head(1)

Entity  Year  Access to electricity (% of population)  \
0  Afghanistan  2000                                 1.613591   

   Access to clean fuels for cooking  \
0                                6.2   

   Renewable-electricity-generating-capacity-per-capita  \
0                                               9.22      

   Financial flows to developing countries (US $)  \
0                                         20000.0   

   Renewable energy share in the total final energy consumption (%)  \
0                                              44.99                  

   Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                 0.16                             0.0   

   Electricity from renewables (TWh)  ...  \
0                               0.31  ...   

   Primary energy consumption per capita (kWh/person)  \
0                                          302.59482    

   Energy intensity level of primary energy (MJ/$2017 PPP GDP)  \
0                                               1.64             

   Value_co2_emissions_kt_by_country  \
0                              760.0   

   Renewables (% equivalent primary energy)  gdp_growth  gdp_per_capita  \
0                                       NaN         NaN             NaN   

   Density\n(P/Km2) Land Area(Km2)  Latitude  Longitude  
0                60       652230.0  33.93911  67.709953  

[1 rows x 21 columns]

In [94]:
df_energy.describe()

Year  Access to electricity (% of population)  \
count  3649.000000                              3639.000000   
mean   2010.038367                                78.933702   
std       6.054228                                30.275541   
min    2000.000000                                 1.252269   
25%    2005.000000                                59.800890   
50%    2010.000000                                98.361570   
75%    2015.000000                               100.000000   
max    2020.000000                               100.000000   

       Access to clean fuels for cooking  \
count                        3480.000000   
mean                           63.255287   
std                            39.043658   
min                             0.000000   
25%                            23.175000   
50%                            83.150000   
75%                           100.000000   
max                           100.000000   

       Renewable-electricity-generating-capacity-per-capita  \
count                                        2718.000000      
mean                                          113.137498      
std                                           244.167256      
min                                             0.000000      
25%                                             3.540000      
50%                                            32.910000      
75%                                           112.210000      
max                                          3060.190000      

       Financial flows to developing countries (US $)  \
count                                    1.560000e+03   
mean                                     9.422400e+07   
std                                      2.981544e+08   
min                                      0.000000e+00   
25%                                      2.600000e+05   
50%                                      5.665000e+06   
75%                                      5.534750e+07   
max                                      5.202310e+09   

       Renewable energy share in the total final energy consumption (%)  \
count                                        3455.000000                  
mean                                           32.638165                  
std                                            29.894901                  
min                                             0.000000                  
25%                                             6.515000                  
50%                                            23.300000                  
75%                                            55.245000                  
max                                            96.040000                  

       Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
count                          3628.000000                     3523.000000   
mean                             70.365003                       13.450190   
std                             348.051866                       73.006623   
min                               0.000000                        0.000000   
25%                               0.290000                        0.000000   
50%                               2.970000                        0.000000   
75%                              26.837500                        0.000000   
max                            5184.130000                      809.410000   

       Electricity from renewables (TWh)  \
count                        3628.000000   
mean                           23.968010   
std                           104.431085   
min                             0.000000   
25%                             0.040000   
50%                             1.470000   
75%                             9.600000   
max                          2184.940000   

       Low-carbon electricity (% electricity)  \
count                             3607.000000   
mean                                36.801182   
std                                 34.31488

In [95]:
len(df_energy.Entity.unique())

176

***
## Modifications de bases

2 modifications de base du dataframe nous semblent nécessaire avant d'aller plus loin.
En effet, un nom de colonne est "Density\n(P/km2)" or le "\n" peut nous poser problème.
Deuxieme changement, cette même densité possède des données de type Object or il serait plus logique que cela soit des float.

In [96]:
df_energy.rename(columns={'Density\\n(P/Km2)': 'Density (P/Km2)'}, inplace=True)

notNa = df_energy["Density (P/Km2)"].dropna()
df_energy["Density (P/Km2)"] = notNa.str.replace(',', '.').astype(float)

In [97]:
df_energy["Density (P/Km2)"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 3649 entries, 0 to 3648
Series name: Density (P/Km2)
Non-Null Count  Dtype  
--------------  -----  
3648 non-null   float64
dtypes: float64(1)
memory usage: 28.6 KB


***
## Analyse détaillée des données

In [98]:
df_energy.isnull().sum()

Entity                                                                 0
Year                                                                   0
Access to electricity (% of population)                               10
Access to clean fuels for cooking                                    169
Renewable-electricity-generating-capacity-per-capita                 931
Financial flows to developing countries (US $)                      2089
Renewable energy share in the total final energy consumption (%)     194
Electricity from fossil fuels (TWh)                                   21
Electricity from nuclear (TWh)                                       126
Electricity from renewables (TWh)                                     21
Low-carbon electricity (% electricity)                                42
Primary energy consumption per capita (kWh/person)                     0
Energy intensity level of primary energy (MJ/$2017 PPP GDP)          207
Value_co2_emissions_kt_by_country                  

In [99]:
df_energy.isnull().sum().sum()

6978

Aucune donnée éssentiel ne manque (le pays ou une année). 
Pour les autres, on cherche à connaitre le type de données manquantes que nous avons : 
- MCAR : 
- MAR : 
- MNAR :

En fonction, nous allons pouvoir adapter la gestion des données manquantes. 

On cherche des valeurs aberrantes

In [100]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [101]:
col = df_energy.columns.tolist()
col.remove('Year')
col.remove('Entity')

In [102]:
# permet de sélectionner le nombre de colones à afficher
cols = col[5:9]
print(cols)

fig = make_subplots(rows=1, cols=len(cols))
for i, var in enumerate(cols):
    fig.add_trace(
        go.Box(y=df_energy[var],
        name=var),
        row=1, 
        col=i+1,
        
            )
fig.update_layout(showlegend=False)
fig.update_traces(boxpoints='outliers', jitter=.3)

['Electricity from fossil fuels (TWh)', 'Electricity from nuclear (TWh)', 'Electricity from renewables (TWh)', 'Low-carbon electricity (% electricity)']


Certaines données paraissent comme aillant beaucoup d'outliers, cependant on peut supposer que certaines ne le sont pas

# Les autres dataset: 


## Datasets

### Pays et continents

In [103]:
df_cc = pd.read_csv(path_cc)

In [104]:
df_cc = df_cc[['country', 'continent', 'sub_region', 'code_3']]
df_cc.head()

country continent       sub_region code_3
0     Afghanistan      Asia    Southern Asia    AFG
1    land Islands    Europe  Northern Europe    ALA
2         Albania    Europe  Southern Europe    ALB
3         Algeria    Africa  Northern Africa    DZA
4  American Samoa   Oceania        Polynesia    ASM

In [105]:
df_cc.isna().sum()

country       0
continent     9
sub_region    9
code_3        0
dtype: int64

In [106]:
df_cc[df_cc['continent'].isna() == True]

country continent sub_region code_3
8                                      Antarctica       NaN        NaN    ATA
30                                  Bouvet Island       NaN        NaN    BVT
32                 British Indian Ocean Territory       NaN        NaN    IOT
46                               Christmas Island       NaN        NaN    CXR
47                        Cocos (Keeling) Islands       NaN        NaN    CCK
78                    French Southern Territories       NaN        NaN    ATF
96              Heard Island and McDonald Islands       NaN        NaN    HMD
206  South Georgia and the South Sandwich Islands       NaN        NaN    SGS
236          United States Minor Outlying Islands       NaN        NaN    UMI

In [107]:
len(df_cc['country'].unique())

249

### IDH (indice de developpement humain)

In [108]:
df_hdi = pd.read_csv(path_hdi)
df_hdi = df_hdi[['Entity', 'Year', 'Human Development Index']]
df_hdi.rename(columns={'Human Development Index': 'HDI'})
df_hdi = df_hdi.query("Year > 1999 & Year < 2021")
df_hdi.tail(4)

Entity  Year  Human Development Index
5918  Zimbabwe  2017                    0.594
5919  Zimbabwe  2018                    0.602
5920  Zimbabwe  2019                    0.601
5921  Zimbabwe  2020                    0.600

### GeoJSON : 
Le geo.json permet d'utiliser Mapbox de plotly. 
Le fichier contient des formes géométriques pour dessiner le contour des frontières.

In [109]:
with open(path_wp, encoding='utf-8') as json_file:
    json_wp = json.load(json_file)
    print(type(json_wp))

<class 'dict'>


In [110]:
json_wp.keys()

dict_keys(['type', 'features'])

In [111]:
for feature in json_wp['features'] : 
    for i, k in enumerate(feature.keys()):
        print(feature[k])
        print(i)
        print()
        
    break

Feature
0

{'featurecla': 'Admin-0 country', 'scalerank': 1, 'labelrank': 5, 'sovereignt': 'Costa Rica', 'sov_a3': 'CRI', 'adm0_dif': 0, 'level': 2, 'type': 'Sovereign country', 'tlc': '1', 'admin': 'Costa Rica', 'adm0_a3': 'CRI', 'geou_dif': 0, 'geounit': 'Costa Rica', 'gu_a3': 'CRI', 'su_dif': 0, 'subunit': 'Costa Rica', 'su_a3': 'CRI', 'brk_diff': 0, 'name': 'Costa Rica', 'name_long': 'Costa Rica', 'brk_a3': 'CRI', 'brk_name': 'Costa Rica', 'brk_group': None, 'abbrev': 'C.R.', 'postal': 'CR', 'formal_en': 'Republic of Costa Rica', 'formal_fr': None, 'name_ciawf': 'Costa Rica', 'note_adm0': None, 'note_brk': None, 'name_sort': 'Costa Rica', 'name_alt': None, 'mapcolor7': 3, 'mapcolor8': 2, 'mapcolor9': 4, 'mapcolor13': 2, 'pop_est': 5047561, 'pop_rank': 13, 'pop_year': 2019, 'gdp_md': 61801, 'gdp_year': 2019, 'economy': '5. Emerging region: G20', 'income_grp': '3. Upper middle income', 'fips_10': 'CS', 'iso_a2': 'CR', 'iso_a2_eh': 'CR', 'iso_a3': 'CRI', 'iso_a3_eh': 'CRI', 'iso_n3': 

In [112]:
i = 0
for feature in json_wp['features']: 
    i+=1
    
print(i)

175


## Comparaison et adaptation : 
Dans cette sectinos on compare les dataset, et notament le nombre de pays

**Nombre de pays par dataset :**

| df_energy | df_cc | json_wp |
|:----------|:------|---------|
| 176       | 249   | 175     |

### Fusions des datasets

On veut concatener les datasets en utilisant les noms de pays comme clée commune. On s'assure donc que tous les noms de pays du dataset de depart (df_energy) sont présent dans les dataset à concatener.

In [113]:
# Pays de df_energy qui ne sont pas dans df_cc

notfound = []
for i in df_energy["Entity"].tolist():
    if (i not in df_cc["country"].tolist()) and (i not in notfound):
        notfound.append(i)
print(notfound)

['Czechia', 'Eswatini', 'North Macedonia', 'United Kingdom', 'United States']


In [114]:
# On modifie les noms des pays de df_cc pour qu'ils correspondent à ceux de df_energy

dic = {"United States of America" : "United States",
       "United Kingdom of Great Britain and Northern Ireland" : "United Kingdom",
       "Macedonia (the former Yugoslav Republic of)" : "North Macedonia",
       "Czech Republic" : "Czechia",
       "Swaziland" : "Eswatini"
       }
df_cc["country"].replace(dic, inplace=True)
df_cc.rename(columns={'country': 'Entity', "continent":"Continent", "sub_region":"Region", "code_3":"iso3"}, inplace=True)

In [115]:
# Pays de df_energy qui ne sont pas dans df_hdi

notfound = []
for i in df_energy["Entity"].tolist():
    if (i not in df_hdi["Entity"].tolist()) and (i not in notfound):
        notfound.append(i)
print(notfound)

['Aruba', 'Bermuda', 'Cayman Islands', 'French Guiana', 'Nauru', 'New Caledonia', 'Puerto Rico', 'Somalia']


Ici les pays qui ne correspondent pas sont en fait des états indépendants faisant partie d'autre pays (ex : French Guiana qui fait partie de la France). Or le dataframe df_hdi ne fait pas la distinction entre les états indépendants et les pays. On va donc volontairement laisser une correspondance vide au niveau de la variable "Entity" pour ces états puisque leurs indices de développement sont inclus dans les data du pays gérant.

In [116]:
# On fusionne les 3 dataframes entre eux
temp = pd.merge(df_energy,df_cc, on = "Entity", how="left")
df_energy = pd.merge(temp,df_hdi, on = ["Entity","Year"], how="left")

In [117]:
df_energy.to_csv(os.path.join("energy-cleaned-dataset.csv"),index=False)